src/main.rs

```diff
@@ -5,60 +5,138 @@ extern crate nalgebra as na;
 mod entity;
 mod velocity_bouncer;
 mod level_reader;
+mod action;
 
 use std::rc::Rc;
 use std::cell::RefCell;
 use piston_window::*;
 use ncollide::world::CollisionWorld;
 use velocity_bouncer::VelocityBouncer;
 use level_reader::LevelReader;
+use action::Action;
+use std::time::SystemTime;
+use entity::{Entity, CollideWorld};
+use na::{Point2, Vector2};
 
-const MU: f32 = 1.0;
+const MU: f32 = 0.99;
 const WIDTH: u32 = 640;
 const HEIGHT: u32 = 480;
+const WAIT_TIME: u32 = 300_000_000; //nanoseconds
 
 fn main() {
     let title = "Goober";
     let mut i = 0;
+    let mut inputs_submitted = false;
+    let mut inputs = Vec::new();
+    let mut timestamp = SystemTime::now();
     let world = Rc::new(RefCell::new(CollisionWorld::new(0.02, true)));
     world.borrow_mut().register_contact_handler("VelocityBouncer", VelocityBouncer);
     let lr = LevelReader::new("levels/level-1.csv");
-    let mut squares = lr.load_level(&world);
+    let squares = lr.load_level(&world);
+    let mut goobs = squares.clone();
+    goobs.retain(|square| square.entity_type == "character");
+    let mut walls = squares.clone();
+    walls.retain(|square| square.entity_type == "wall");
     let mut window: PistonWindow = WindowSettings::new(title, [WIDTH, HEIGHT])
         .exit_on_esc(true)
         .build()
         .unwrap_or_else(|e| panic!("Failed to build PistonWindow: {}", e));
+    let font = "assets/FiraSans-Regular.ttf";
+    let factory = window.factory.clone();
+    let mut glyphs = Glyphs::new(font, factory).unwrap();
+    let sprite_path = "./assets/green-blob-hi.png";
+    let sprite = Texture::from_path(
+            &mut window.factory,
+            &sprite_path,
+            Flip::None,
+            &TextureSettings::new())
+            .unwrap();
 
     while let Some(e) = window.next() {
 
         if let Some(_) = e.render_args() {
             window.draw_2d(&e, |c, g| {
                 clear([0.0, 0.0, 0.0, 1.0], g);
-                for sq in &squares {
-                    rectangle(sq.color(), sq.geometry(), c.transform, g);
+                for wall in &walls {
```

Comment:
I agree. It seems tricky, though, as that would mean passing the gfx_graphics::back_end::GfxGraphics (`g`) into a function on Entity, but the `back_end` module is private, so it can't be included directly. Since not all rendering involves the entity (like clear and drawing the command text), I don't have a clear idea of how to pull that off yet.